## Programmatic Assessments in pandas
### Gather

In [1]:
import pandas as pd

In [2]:
patients = pd.read_csv('patients.csv')
treatments = pd.read_csv('treatments.csv')
adverse_reactions = pd.read_csv('adverse_reactions.csv')

### Assess
These are the programmatic assessment methods in pandas that you will probably use most often:

* .head (DataFrame and Series)
* .tail (DataFrame and Series)
* .sample (DataFrame and Series)
* .info (DataFrame only)
* .describe (DataFrame and Series)
* .value_counts (Series only)
* Various methods of indexing and selecting data (.loc and bracket notation with/without boolean indexing, also .iloc)

Try `.head` and `.tail` on the `patients` table.

In [5]:
patients.head(3)

patient_id assigned_sex given_name  surname                    address  \
0           1       female        Zoe  Wellish       576 Brown Bear Drive   
1           2       female     Pamela     Hill  2370 University Hill Road   
2           3         male        Jae   Debord      1493 Poling Farm Road   

                city       state  zip_code        country  \
0  Rancho California  California   92390.0  United States   
1          Armstrong    Illinois   61812.0  United States   
2               York    Nebraska   68467.0  United States   

                                 contact  birthdate  weight  height   bmi  
0   951-719-9170ZoeWellish@superrito.com  7/10/1976   121.7      66  19.6  
1  PamelaSHill@cuvox.de+1 (217) 569-3204   4/3/1967   118.8      66  19.2  
2       402-363-6804JaeMDebord@gustr.com  2/19/1980   177.8      71  24.8

In [6]:
patients.tail(3)

patient_id assigned_sex given_name      surname  \
500         501       female      Jinke    de Keizer   
501         502       female    Chidalu  Onyekaozulu   
502         503         male        Pat      Gersten   

                      address           city     state  zip_code  \
500         649 Nutter Street  Overland Park        MO   64110.0   
501  3652 Boone Crockett Lane        Seattle        WA   98109.0   
502         2778 North Avenue           Burr  Nebraska   68324.0   

           country                                          contact  \
500  United States            816-223-6007JinkedeKeizer@teleworm.us   
501  United States  ChidaluOnyekaozulu@jourrapide.com1 360 443 2060   
502  United States             PatrickGersten@rhyta.com402-848-4923   

     birthdate  weight  height   bmi  
500  1/13/1971   171.2      67  26.8  
501  2/13/1952   176.9      67  27.7  
502   5/3/1954   138.2      71  19.3

In [16]:
patients.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 503 entries, 0 to 502
Data columns (total 14 columns):
patient_id      503 non-null int64
assigned_sex    503 non-null object
given_name      503 non-null object
surname         503 non-null object
address         491 non-null object
city            491 non-null object
state           491 non-null object
zip_code        491 non-null float64
country         491 non-null object
contact         491 non-null object
birthdate       503 non-null object
weight          503 non-null float64
height          503 non-null int64
bmi             503 non-null float64
dtypes: float64(3), int64(2), object(9)
memory usage: 55.1+ KB


In [19]:
patients.describe()

patient_id      zip_code      weight      height         bmi
count  503.000000    491.000000  503.000000  503.000000  503.000000
mean   252.000000  49084.118126  173.434990   66.634195   27.483897
std    145.347859  30265.807442   33.916741    4.411297    5.276438
min      1.000000   1002.000000   48.800000   27.000000   17.100000
25%    126.500000  21920.500000  149.300000   63.000000   23.300000
50%    252.000000  48057.000000  175.300000   67.000000   27.200000
75%    377.500000  75679.000000  199.500000   70.000000   31.750000
max    503.000000  99701.000000  255.900000   79.000000   37.700000

In [23]:
patients.sample(5)

patient_id assigned_sex  given_name   surname                   address  \
24           25         male       Jakob  Jakobsen         648 Old Dear Lane   
455         456       female         Hui    Ch'eng  3423 Burning Memory Lane   
365         366         male      Robert    Wagner       162 Brooklyn Street   
12           13         male  Omeokachie  Ibeamaka        2544 Worley Avenue   
394         395       female        Anja   Mueller          570 Alpha Avenue   

            city     state  zip_code        country  \
24   Port Jervis  New York   12771.0  United States   
455      Yardley        PA   19067.0  United States   
365        Salem        VA   24153.0  United States   
12     Lynchburg        VA   24504.0  United States   
394        Tyler        TX   75757.0  United States   

                                        contact  birthdate  weight  height  \
24   JakobCJakobsen@einrot.com+1 (845) 858-7707   8/1/1985   155.8      67   
455              215-321-9611HuiCheng@rhyta.com   5/3/1982   203.1      69   
365     RobertWagner@jourrapide.com540-986-1369  6/20/1925   200.0      67   
12    OmeokachieIbeamaka@einrot.com434-509-2614   8/5/1957   224.2      69   
394           903-939-1025AnjaMueller@gustr.com   8/6/1942   182.2      61   

      bmi  
24   24.4  
455  30.0  
365  31.3  
12   33.1  
394  34.4

In [21]:
patients[patients['city'] == 'New York'].shape

(18, 14)

In [18]:
patients[patients['address'].isnull()]

patient_id assigned_sex given_name      surname address city state  \
209         210       female     Lalita  Eldarkhanov     NaN  NaN   NaN   
219         220         male         Mỹ        Quynh     NaN  NaN   NaN   
230         231       female  Elisabeth      Knudsen     NaN  NaN   NaN   
234         235       female    Martina    Tománková     NaN  NaN   NaN   
242         243         male       John      O'Brian     NaN  NaN   NaN   
249         250         male   Benjamin       Mehler     NaN  NaN   NaN   
257         258         male        Jin         Kung     NaN  NaN   NaN   
264         265       female   Wafiyyah       Asfour     NaN  NaN   NaN   
269         270       female     Flavia   Fiorentino     NaN  NaN   NaN   
278         279       female   Generosa        Cabán     NaN  NaN   NaN   
286         287         male      Lewis         Webb     NaN  NaN   NaN   
296         297       female        Chỉ          Lâm     NaN  NaN   NaN   

     zip_code country contact   birthdate  weight  height   bmi  
209       NaN     NaN     NaN   8/14/1950   143.4      62  26.2  
219       NaN     NaN     NaN    4/9/1978   237.8      69  35.1  
230       NaN     NaN     NaN   9/23/1976   165.9      63  29.4  
234       NaN     NaN     NaN    4/7/1936   199.5      65  33.2  
242       NaN     NaN     NaN   2/25/1957   205.3      74  26.4  
249       NaN     NaN     NaN  10/30/1951   146.5      69  21.6  
257       NaN     NaN     NaN   5/17/1995   231.7      69  34.2  
264       NaN     NaN     NaN   11/3/1989   158.6      63  28.1  
269       NaN     NaN     NaN   10/9/1937   175.2      61  33.1  
278       NaN     NaN     NaN  12/16/1962   124.3      69  18.4  
286       NaN     NaN     NaN    4/1/1979   155.3      68  23.6  
296       NaN     NaN     NaN   5/14/1990   181.1      63  32.1

Try `.sample` on the `treatments` table.

In [7]:
treatments.sample(3)

given_name      surname    auralin    novodra  hba1c_start  hba1c_end  \
198     amalie  christensen  38u - 45u          -         7.88       7.55   
206        emy       markus          -  29u - 24u         7.53       7.14   
95      jaakko    eskelinen  39u - 45u          -         7.73       7.42   

     hba1c_change  
198          0.33  
206          0.39  
95           0.31

Try `.info` on the `treatments` table.

In [8]:
treatments.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 280 entries, 0 to 279
Data columns (total 7 columns):
given_name      280 non-null object
surname         280 non-null object
auralin         280 non-null object
novodra         280 non-null object
hba1c_start     280 non-null float64
hba1c_end       280 non-null float64
hba1c_change    171 non-null float64
dtypes: float64(3), object(4)
memory usage: 15.4+ KB


Try `.describe` on the `patients` table.

In [20]:
treatments.describe()

hba1c_start   hba1c_end  hba1c_change
count   280.000000  280.000000    171.000000
mean      7.985929    7.589286      0.546023
std       0.568638    0.569672      0.279555
min       7.500000    7.010000      0.200000
25%       7.660000    7.270000      0.340000
50%       7.800000    7.420000      0.380000
75%       7.970000    7.570000      0.920000
max       9.950000    9.580000      0.990000

Try `.value_counts` on the *adverse_reaction* column of the `adverse_reactions` table.

In [17]:
adverse_reactions.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 34 entries, 0 to 33
Data columns (total 3 columns):
given_name          34 non-null object
surname             34 non-null object
adverse_reaction    34 non-null object
dtypes: object(3)
memory usage: 896.0+ bytes


In [22]:
adverse_reactions.describe()

given_name  surname adverse_reaction
count          34       34               34
unique         34       33                6
top        joseph  johnson     hypoglycemia
freq            1        2               19

In [10]:
adverse_reactions.adverse_reaction.value_counts()

hypoglycemia                 19
injection site discomfort     6
headache                      3
cough                         2
throat irritation             2
nausea                        2
Name: adverse_reaction, dtype: int64

Try selecting the records in the `patients` table for patients that are from the *city* New York.

#### Quality

##### patients table

* zip code is a float not a strig
* for some registries zip code has four digits
* Tim Neudorf height is 27 instead of 71 inches
* Sometimes there's full state name and other times just the abbreviations
* patient_id 9 has name with typo: Dsvid Gustafson
* missing demographic information (address - contact columns)
* there are 4 columns with wrong data types: assigned_sex, state, zip_code and birthdate)
* different phone number formats

##### treatements table

* missing HbA1c changes
* the letter u in starting and ending doses for Auralin and Novodra
* lowercase given names and surnames
* missing records (280 instead of 350)
* wrong data types for auralin and novodra columns
* inaccurate value for HbA1c changes (4s mistaken by 9s)

##### adverse_reactions table

* lowercase given names and surnames